Dfe.Data.SearchPrototype/Web/Tests/PageIntegrationTests/SearchResultsPageTests.cs

```diff
@@ -109,15 +111,24 @@ public async Task Search_ByName_ReturnsMultipleResults()
                     ["searchKeyWord"] = "Academy"
                 });
 
-            _logger.WriteLine("SendAsync client base address: " + _client.BaseAddress);
-            _logger.WriteLine("SendAsync request message: " + formResponse.RequestMessage!.ToString());
-
             var resultsPage = await HtmlHelpers.GetDocumentAsync(formResponse);
 
-            _logger.WriteLine("Document: " + resultsPage.Body!.OuterHtml);
-
-            resultsPage.QuerySelector(SearchPage.SearchResultsNumber.Criteria)!.TextContent.Should().Contain("Results");
-            resultsPage.GetMultipleElements(SearchPage.SearchResultLinks.Criteria).Count().Should().BeGreaterThan(1);
+            resultsPage.GetElementText(SearchResultsPage.SearchResultsNumber.Criteria).Should().Contain("Results");
+            resultsPage.GetMultipleElements(SearchResultsPage.SearchResultLinks.Criteria).Count().Should().Be(2);
+
+            resultsPage.GetElementText(SearchResultsPage.SearchResultEstablishmentName(Constants.Urns.GOOSE_ACADEMY).Criteria).Should().Be("Goose Academy");
+            resultsPage.GetElementText(SearchResultsPage.SearchResultEstablishmentUrn(Constants.Urns.GOOSE_ACADEMY).Criteria).Should().Contain("123456");
+            resultsPage.GetElementText(SearchResultsPage.SearchResultEstablishmentAddress(Constants.Urns.GOOSE_ACADEMY).Criteria).Should().Contain("Goose Street, Goose Locality, Goose Address 3, Goose Town, GOO OSE");
+            resultsPage.GetElementText(SearchResultsPage.SearchResultEstablishmentType(Constants.Urns.GOOSE_ACADEMY).Criteria).Should().Contain("Academy");
+            resultsPage.GetElementText(SearchResultsPage.SearchResultEstablishmentStatus(Constants.Urns.GOOSE_ACADEMY).Criteria).Should().Contain("Open");
+            resultsPage.GetElementText(SearchResultsPage.SearchResultEstablishmentPhase(Constants.Urns.GOOSE_ACADEMY).Criteria).Should().Contain("Secondary");
+                                                         
+            resultsPage.GetElementText(SearchResultsPage.SearchResultEstablishmentName(Constants.Urns.HORSE_ACADEMY).Criteria).Should().Be("Horse Academy");
+            resultsPage.GetElementText(SearchResultsPage.SearchResultEstablishmentUrn(Constants.Urns.HORSE_ACADEMY).Criteria).Should().Contain("234567");
+            resultsPage.GetElementText(SearchResultsPage.SearchResultEstablishmentAddress(Constants.Urns.HORSE_ACADEMY).Criteria).Should().Contain("Horse Street, Horse Locality, Horse Address 3, Horse Town, HOR SEE");
+            resultsPage.GetElementText(SearchResultsPage.SearchResultEstablishmentType(Constants.Urns.HORSE_ACADEMY).Criteria).Should().Contain("Academy");
+            resultsPage.GetElementText(SearchResultsPage.SearchResultEstablishmentStatus(Constants.Urns.HORSE_ACADEMY).Criteria).Should().Contain("Open");
+            resultsPage.GetElementText(SearchResultsPage.SearchResultEstablishmentPhase(Constants.Urns.HORSE_ACADEMY).Criteria).Should().Contain("Post 16");
```

Comment:
I think we'd all benefit from revisiting the kind of Page model that we've looked at in the past - it would help to make these tests (and our dev tests)  more efficient because we could get the results part of the page, and drill down into the bits we want, rather than here - traversing the whole DOM for each assert. Asia added a container for the search results bit, and then we want to be able to easily get specific bits from within that container.
Not for this PR though

Reply:
Yep been talking to @aahmed-dfe about this. He's implementing a new `IPageQueryClient` in CSCP and soon into our common framework to clean this up a bit.

Reply:
Hello! I've implemented part of this in CSCP.

@CathLass  @antk43  some description [in the PR here for how I've changed CSCP PageModels](https://agilefactory.visualstudio.com/SPT/_git/Web/pullrequest/5348). I agree with the PageModel needing revisiting so it's more reuseable across different types of tests.

I think the root of the problem is we're exposing coupled Selenium types through `IWebElement` and `IWebDriver` in our page constructors and properties. That dependency needs inverting and the surface of the PageModel shouldn't be aware of the library we use to query the DOM, whether than be AngleSharp, Selenium, PlayWright etc.

I've added an abstraction so that each page takes `IPageQueryClient` in. Which looks roughly like below.

Then different types of tests can provide their query mechanism

```cs
public interface IPageQueryClient
{
    string? GetText(string cssSelector);
    string? GetAttribute(string cssSelector, string attribute);
// ....
}


```


```
cs
builder.ConfigureServices(t => t.AddSingleton<IPageQueryClient, AngleSharpQueryClient>();
```


```cs
[Test]
public async Task UITestThatChecksHeading(){

IPageQueryClient pageClient = GetClient();
new HomePage(pageClient).GetHeading().Should().Be("something");

}
```

```cs
// TEST
[Test]
public async Task HTTPTestThatChecksHeading(){
IPageQueryClient client = GetClient();
new HomePage(client).GetHeadingText().Should().Be("TEST");

}
```

Reply:
The reuse then happens at the exposed `PageModel.Method`.

I am not fully sold on t he abstracting, but I'd thought this is a good place to start decoupling the layers.

Reply:
Looks neat, could we register the page objects and have the IPageQueryClient materialised that way rather than extracting from the container, instantiating the page object, etc

Reply:
```
because we could get the results part of the page, and drill down into the bits we want, rather than here 
- traversing the whole DOM for each assert
```

Not sure this answers @CathLass question though about making the querying more efficient.

The speed of the queries over the DOM are usually very fast through AngleSharp, after it's parsed the document - we're talking milliseconds for most queries - be good to get some numbers though. 

*I think* most of the work happens in yielding a document - not the query over the doc.

-  `WebApplicationFactory.CreateClient().GetAsync(route)` which hosts the app and goes through the application layer.
- AngleSharps parsing the response back and mapping it through `HtmlHelpers.GetDocumentAsync(HttpResponseMessage)`.

You could use more specific cssSelectors like `#search-container .my-thing` but the query would still operate over the parsed document?

Reply:
Another thing is you could segregate the PageModel into different parts of the page, but that wouldn't increase speed, just make it easier to separate and componentise part of the page;

```cs

public sealed Page : BasePage {

public class Page(IPageQueryClient pageQueryClient){
  MyComponent = new(pageQueryClient);
}

public Component MyComponent { get; }
}


public sealed class Component : BasePage{
  private static string MyLocator => "#some-locator";
}
```

Then in your tests

`Page(client).Component.GetFoo().Should().Be("bar");

Reply:
We had something like this a few months ago, then we were composing the page objects using more atomic page component parts to promote re-use across pages (we decided the scale of this prototype didn't warrant this). On a larger system we could create reusable page object components (e.g. for tabular data, etc) which we could then utilise at the page object level. Anyway, food for thought going forwards.

Reply:
Obvs as suggested, this won't benefit us in terms of speeding up traversal of the DOM, rather allows us to rapidly prototype new page objects when required

Reply:
> Looks neat, could we register the page objects and have the IPageQueryClient materialised that way rather than extracting from the container, instantiating the page object, etc

Yeah that'd be ideal, just register everything into the container and resolve it via the constructor of the test.

A constraint I've hit with AngleSharp is that sometimes tests build up runtime-known dependencies. Such as `school/{urn}/{name}` and we only discover the name of the school at runtime.

So I think we'll widen what the constructors of these QueryClients accept to let it resolve through the container, or possibly put a factory in front to do the client creation.... I'm not sure yet :)

```cs

public sealed class AngleSharpQueryClient : IPageQueryClient{

public AngleSharpQueryClient(IHtmlDocument doc) // <<< can't resolve this til runtime so needs to get the document in a delayed manner.
}

```

Reply:
Yeah sounds good, I was thinking factory but early days I guess ...top work though dude. We defo need to get together as a collective and discuss strategies/options etc. I know Cath and Roger have pushed for this 👍